src/tiny_gea2_single_wire_interface.c

```diff
@@ -0,0 +1,647 @@
+typedef struct {
+  uint8_t destination;
+  uint8_t data_length;
+  uint8_t source;
+  uint8_t data[1];
+} send_packet_t;
+
+enum {
+  signal_byte_received = tiny_fsm_signal_user_start,
+  signal_interbyte_timeout,
+  signal_send_ready,
+  signal_idle_cooldown_timeout,
+  signal_reflection_timeout,
+  signal_collision_idle_timeout,
+  signal_ack_timeout
+};
+
+enum {
+  send_packet_header_size = offsetof(send_packet_t, data),
+  data_length_bytes_not_included_in_data = tiny_gea3_packet_transmission_overhead - tiny_gea3_packet_overhead,
+  crc_size = sizeof(uint16_t),
+  packet_bytes_not_included_in_payload = crc_size + offsetof(tiny_gea3_packet_t, payload),
+  unbuffered_bytes = 2 // STX, ETX
+};
+
+enum {
+  send_state_data,
+  send_state_crc_msb,
+  send_state_crc_lsb,
+  send_state_etx,
+  send_state_stx,
+  send_state_done,
+};
+
+static void state_idle(tiny_fsm_t* fsm, const tiny_fsm_signal_t signal, const void* data);
+static void state_receive(tiny_fsm_t* fsm, const tiny_fsm_signal_t signal, const void* data);
+static void state_idle_cooldown(tiny_fsm_t* fsm, const tiny_fsm_signal_t signal, const void* data);
+static void state_send(tiny_fsm_t* fsm, const tiny_fsm_signal_t signal, const void* data);
+static void state_wait_for_ack(tiny_fsm_t* fsm, const tiny_fsm_signal_t signal, const void* data);
+static void state_collision_cooldown(tiny_fsm_t* fsm, const tiny_fsm_signal_t signal, const void* data);
+
+static self_t* interface_from_fsm(tiny_fsm_t* fsm)
+{
+  return container_of(self_t, _private.fsm, fsm);
+}
+
+static void byte_received(void* context, const void* _args)
+{
+  reinterpret(instance, context, self_t*);
+  reinterpret(args, _args, const tiny_uart_on_receive_args_t*);
+
+  tiny_fsm_send_signal(&instance->_private.fsm, signal_byte_received, &args->byte);
+}
+
+#define needs_escape(_byte) ((_byte & 0xFC) == tiny_gea3_esc)
+
+#define is_broadcast_address(_address) ((gea2_broadcast_mask & _address) == gea2_broadcast_mask)
+
+static void state_idle(tiny_fsm_t* fsm, const tiny_fsm_signal_t signal, const void* data)
+{
+  self_t* instance = interface_from_fsm(fsm);
+  switch(signal) {
+    case tiny_fsm_signal_entry:
+    case signal_send_ready:
+      if(instance->_private.send.active) {
+        tiny_fsm_transition(fsm, state_send);
+      }
+      break;
+
+    case signal_byte_received: {
+      reinterpret(byte, data, const uint8_t*);
+
+      if(instance->_private.receive.packet_ready) {
+        break;
+      }
+
+      if(*byte == tiny_gea3_stx && !instance->_private.receive.packet_ready) {
+        tiny_fsm_transition(fsm, state_receive);
+      }
+      else {
+        tiny_fsm_transition(fsm, state_idle_cooldown);
+      }
+    } break;
+  }
+}
+
+static void reflection_timeout(void* context)
+{
+  self_t* instance = context;
+  tiny_fsm_send_signal(&instance->_private.fsm, signal_reflection_timeout, NULL);
+}
+
+static bool determine_byte_to_send_considering_escapes(self_t* instance, uint8_t byte, uint8_t* byteTosend)
```

Comment:
The `escape` byte is the final byte of the protocol. It basically says `the message is done now`

Reply:
Actually, `escape` is how we distinguish between control characters acting as control characters (for example, start of frame, end of frame) and data that happens to look like control characters. So as an example, if we have a byte that matches STX in the middle of the payload, we'll prefix it with an ESC byte so that the receiver knows it's not really an STX control character, it's data.